### PR TITLE
client: use a separate ctx for each request

### LINKED ISSUE #2849: member list doesn't respond to non-full cluster ("context deadline exceeded")
I specified a five-node cluster using the discovery service.  Four nodes are started and visible to each other, as verified in the journal.  However, no node can view the existing member list.

core@vb-command ~ $ etcdctl --debug member list
context deadline exceeded
core@vb-command ~ $ etcdctl --version
etcdctl version 2.0.10

On all nodes: etcdctl ls blocks until all five nodes are in the cluster.  Is this documented anywhere?  This design is not very flexible.  A 4-active-node cluster should be fully functional, even with a non-existing fifth (presumably failed) node.

### PR DESCRIPTION
**Pull request by xiang90**
ClusterClient should use a separate ctx for each request. Or
if one of the endpoint is dead, it might use up the full timeout
duration. So we have no chance to switch to another endpoint.

/cc @bcwaldon 

This is still WIP. I need more test around it and clean up the code a little bit. 
But I would like to see your opinion on this. 

Another simpler solution is to assume the given `ctx` is per request not per the
whole operation. So if we have 5 endpoints, if all of them timed out the operation
might consume 5 * ctx.Timeout. This might break the API contract, but is more clear
in the implementation point of view.

How do you think?

Fix #2849

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -191,17 +191,31 @@ func (c *httpClusterClient) Do(ctx context.Context, act httpAction) (*http.Respo
...
+		epctx    context.Context
+		epcancel context.CancelFunc
+	)
 
 	for _, ep := range eps {
+		deadline, ok := ctx.Deadline()
+		if ok {
+			epctx, epcancel = context.WithTimeout(ctx, time.Duration(int(deadline.Sub(time.Now()))/len(eps)))
+		} else {
+			epctx = epctx
```
Comment (bcwaldon) [inline client/client.go]: What is this assignment achieving?

Comment (xiang90) [inline client/client.go]: yes... this shit is not tested... want to know if this is the approach you like.

**Comment by bcwaldon** [comment]
@xiang90 I understand the motivation. Maybe it would be cleaner to simply tie the transport Timeout to `ctx.Deadline() - time.Now()`?

**Comment by xiang90** [comment]
@bcwaldon Hmm... 

The first thing I want to do is to evenly divide the timeouts to possible endpoints. So no single endpoint can use up all the time out.

The transport timeout is actually the next thing I want to fix. It is currently longer than the ctx timeout (the defaults), which makes no sense.

File: client/client_test.go [hunk 2]
```diff
@@ -328,7 +328,7 @@ func TestHTTPClusterClientDo(t *testing.T) {
 		// context.DeadlineExceeded short-circuits Do
 		{
 			client: &httpClusterClient{
-				endpoints: []url.URL{fakeURL, fakeURL},
+				endpoints: []url.URL{fakeURL},
```
Comment (xiang90) [inline client/client_test.go]: The previous test is not correct after this commit. If it has two endpoints, we should expect it try the second one before returning error. Use single endpoint to fix the test.

File: client/client_test.go [hunk 3]
```diff
@@ -412,6 +412,53 @@ func TestHTTPClusterClientDo(t *testing.T) {
...
+	}
+
+	neps := len(c.endpoints)
+
+	go func() {
+		// feed the chan since it has one buffer
+		tr.finishCancel <- struct{}{}
+
+		// expect "neps" timeout cancel
+		start := time.Now()
```
Comment (yichengq) [inline client/client_test.go]: put the start into the loop?

File: client/client_test.go [hunk 4]
```diff
@@ -412,6 +412,53 @@ func TestHTTPClusterClientDo(t *testing.T) {
...
+		for i := 0; i < neps; i++ {
+			tr.finishCancel <- struct{}{}
+			d := time.Since(start)
+			if d > timeout/time.Duration(neps)+10*time.Millisecond || d < timeout/time.Duration(neps)-10*time.Millisecond {
+				t.Errorf("#%d: timeout = %v, want around %v", i, d, timeout/3)
```
Comment (yichengq) [inline client/client_test.go]: s/3/time.Duration(neps)

**Comment by yichengq** [comment]
Could we record this behavior at the comment? the users of client package should know it and prepare for it. For example, if one request could be finished in 1s and it has 10 endpoints, the user may give 5s-timeout context because 5s > 1s, which leads to the result that he always fails to do the request.

**Comment by xiang90** [comment]
@yichengq I do not quite understand your example.

**Comment by yichengq** [comment]
It should eliminate the possible user expectation that the timeout in context could be used for one single request.

Say one specific request can be finished in 1s. The user may give a context without 5s timeout, and think that is enough for it to be finished. But that is not true sometimes. If the clusterClient has 10 endpoints, it gives each endpoint only 0.5s to do the request, which is not enough for the user.

**Comment by xiang90** [comment]
> Say one specific request can be finished in 1s.

Ah, I guess what you want to say is: a request cannot be finished in 1 second, right? Or it still does not make a lot sense to me.

**Comment by yichengq** [comment]
@xiang90 Yes.

@bcwaldon After discussion with @xiang90 , adding the fact that we divide the context timeout into each endpoint is bad because 1. users have to understand the concept of endpoints and use it when creating context 2. it limits further optimization, e.g., bypass the endpoints of the same members

Here are two solutions for the original issue that come to my head:

1. keep the original API and ask users to set transport.Timeout

Users have to understand that we try endpoint one by one, and may fail if it hasn't tried the good one. They could set shorter transport.Timeout to go through all endpoints, but that may give them the risk that the time is not enough for one single request to finish.

2. use timeout instead of context

This makes it work as standard http package.

**Comment by xiang90** [comment]
> we divide the context timeout

I do not think this is "bad" at all. I think this can work. I want to start with something simple. This timeout thing is a hard problem. I have many thoughts into this and I do not think we can make it work perfectly any time soon.

**Comment by yichengq** [comment]
@xiang90 I wrote it wrong: s/adding the fact/adding the fact into the doc/

**Comment by xiang90** [comment]
> use timeout instead of context

They are not exclusive at all. We can add timeout for different phase and different endpoints. Context can control cancel and the overall deadline for you.

**Comment by yichengq** [comment]
> They are not exclusive at all. We can add timeout for different phase and different endpoints. Context can control cancel and the overall deadline for you.

Ack. context has its own usage and should not be discarded.

**Comment by xiang90** [comment]
@yichengq Since you would like to take care of this issue, I am closing this pull request.

**Pull request closed without merging**